File: Java/프로그래머스/2/12923. 숫자 블록/숫자 블록.java
class Solution {
    public int[] solution(long begin, long end) {
        int[] result = new int[(int) (end - begin + 1)];
        int index = 0;
        for (long i = begin; i <= end; i++, index++) {
            result[index] = getMaxDivisor((int) i);
        }

        return result;
    }

    private int getMaxDivisor(int number) {
        if (number <= 1) {
            return 0;
        }

        int result = 1;

        for (int i = 2; i <= Math.sqrt(number); i++) {
            if (number % i == 0) {
                if (number / i <= 10000000) {
                    return number / i;
                }

                if (i <= 10000000) {
                    result = Math.max(result, i);
                } else {
                    break;
                }
            }
        }

        return result;
    }
}

File: Python/프로그래머스/2/12923. 숫자 블록/숫자 블록.py
def solution(begin, end):
    return [get_max_divisor(i) for i in range(begin, end + 1)]


def get_max_divisor(n):
    if n == 1:
        return 0

    arr = [1]
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            arr.append(i)
            if n // i <= 10_000_000:
                arr.append(n // i)

    return max(arr)